### ISSUE TITLE
questions about the codes.

### ISSUE THREAD

**Issue by GANNNN123** [lmnt-com/diffwave#1]
Thanks for your great work!
I study with your codes and It is of great help to me.

By the way, I have two questions in your scripts.

1.
In learner.py scripts, The variable 'noise_level'(line 50) takes square root, and it seems to mean 'alpha_cumprod_sqrt' in the paper.
but in line120&121, it takes square_root again and the variable 'noise_scale_sqrt' seems to be took square root second time.
(+'noise_scale' seems to equal to 'alpha_cumprod_sqrt' in paper.)
I thought the input 'noisy_audio' is different from the original paper.

To confirm whether there is something wrong, I trained both model with&without changing the script(remove **0.5 in line 50).
and I found the model without changing the code(same with your scripts) acts better which means your code is right!

I have checked the scripts several times, but I still find there is something weird in the scripts(alpha_cumprod took sqrt two times)
so I cannot understand the given results.

May I ask you to confirm whether your code is exactly same with the paper and if I missed anything?

2. 
You used L1 distance for the objective function but originally, the loss is euclidian distance in the paper.
Is there any reason for using L1 distance? (I knew that wavegrad paper said training with L1 is better!)

Thanks for your work again. :)

**Comment by GANNNN123** [comment]
I said I have checked both model with&without changing the code, but the training took few steps.
So I didn't checked perfectly trained samples, the result can be wrong. Please check whether there is something wrong.

**Comment by sharvil** [comment]
Thanks for taking an interest in this project, @GANNNN123.

1. That's an excellent catch. Thanks for reviewing the code! You're right, it's a bug in the implementation and differs from the paper. The training code has this bug but the inference code does not. You're seeing good behavior with this implementation early in training because the extra sqrt increases the noise level during training but at inference time we inject smaller noise. I'll send out a bug fix for this soon.

2. For waveform reconstruction loss, I get better results with L1 loss versus L2 loss (L2 tends to over-smooth). In this case, there's no compelling reason to choose L2 over L1 so I went with L1 as a good default.

**Comment by GANNNN123** [comment]
Thank you for answering the questions! As you mentioned, I checked the fixed one has better result after training about 24 hours!

**Comment by sharvil** [comment]
Perfect – I'm glad it worked for you. Thanks for confirming the fix; I'll close the issue now.
